Modules/atexitmodule.c

```diff
@@ -114,6 +131,7 @@ _PyAtExit_Fini(PyInterpreterState *interp)
         PyMem_Free(callback);
         exitfunc(data);
     }
+    _PyAtExit_UNLOCK(state);
```

Comment:
I would prefer to unlock *before* the loop. `exitfunc()` may create a reentrant call. *state* is no longer accessed in the loop.

Reply:
I think we just need to unlock for `exitfunc`. It will be racy to access callbacks without the lock.